src/System.Net.Security/tests/FunctionalTests/SslStreamAlpnTests.cs

```diff
@@ -140,7 +139,7 @@ public void SslStream_StreamToStream_Alpn_NonMatchingProtocols_Fail()
                     ServerCertificate = certificate,
                 };
                 
-                Task t1 = Assert.ThrowsAsync<AuthenticationException>(() => client.AuthenticateAsClientAsync(clientOptions, CancellationToken.None));
+                Task t1 = Assert.ThrowsAsync<TimeoutException>(() => client.AuthenticateAsClientAsync(clientOptions, CancellationToken.None));
```

Comment:
TimeoutException is correct here?  I thought from your description that it wasn't.

Reply:
The timeout exception is correct, because the server doesn't send a failure notice to client, so client just waits for reply from server about next handshake status. Since it doesn't receive anything it times out.

Reply:
> the server doesn't send a failure notice to client

This particular server, or in general?  If just this particular server, how can we test that the right exception occurs when the server does send a failure notice?

Reply:
The sslstream server implementation. The schannel or openssl has to return an output buffer containing alert information to be sent regarding why handshake failed. On debugging I don't see the outgoing buffers having any information regarding the alert for applicationprotocolmismatch. So the sslstream as server doesn't write anything to wire to indicate failure, it just throws AuthenticationException and disposes the connection.

Reply:
So, if anyone is using sslstream as server, the handshake process will not send alert for alpn failure. If they use sslstream as client and someoother server that supports sending alerts, the client will receive the alert and throw authenticationexception as well.

Reply:
If an exception is thrown out of the authenticate method then shouldn't the test end immediately. I would say either the code isn't exiting straight away or the test is incorrect. A timeout for 20secs (from memory the time in the virtual network stream) doesn't seem to be the correct behaviour either way.

Reply:
> The sslstream server implementation

Sounds like something we need to fix.

> the client will receive the alert and throw authenticationexception as well.

We should have a separate test for that if we don't already.

> If an exception is thrown out of the authenticate method then shouldn't the test end immediately

Yeah, it sounds like the test is buggy regardless.

Reply:
> If an exception is thrown out of the authenticate method then shouldn't the test end immediately

It does, the except is thrown out of Authenticate method for sslstream as server, and the check that for AuthenticationException on server sslstream passes. But the Authenticate method as client, doesn't throw exception, because it doesn't receive the alpn failure alert or the close_notify alert, so according to client, it is still waiting for handshake data, and it timesout on read wait during handshake

Reply:
Then in the test can you set the server ssl stream to own the network stream so it closes it on error then the client side should terminate straightaway

Reply:
> Then in the test can you set the server ssl stream to own the network stream so it closes it on error then the client side should terminate straightaway

It does own the internal stream as well, but this is not an actual network stream. If you look at the VirtualNetworkStream, it just has a concurrentqueue for client/server from which it reads/writes to based on semaphore signalling. The actual network stream tests are in SslStreamNetworkStreamTest.cs, and it uses a Tcpclient/listener for connections. If the timeout is the problem here, I can move the test to be an actual network stream test.

Reply:
I guess my concern is not today as you've just worked on it but in the future if a change is made the test won't say if a real hang has been introduced or not.